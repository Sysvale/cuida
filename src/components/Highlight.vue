<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<template
		v-if="highlightedText && isAValidSubstring"
	>
		<template
			v-for="(element) in substrings"
			:key="element.id"
		>
			<mark
				v-if="element.highlighted"
				class="highlight__container"
				:class="dynamicHighlightClass"
				:style="dynamicStyle"
			>
				{{ element.string }}
			</mark>
			<template v-else>
				{{ element.string }}
			</template>
		</template>
	</template>

	<mark
		v-else
		class="highlight__container"
		:class="dynamicHighlightClass"
		:style="dynamicStyle"
	>
		<!-- @slot Slot usado para especificar o texto que receberá o highlight. -->
		<slot />
	</mark>
</template>

<script>
export default {
	props: {
		/**
		* A variante do Highlight. São 3 variantes implementadas: 'info', 'success'
		* e 'danger'
		*/
		variant: {
			type: String,
			default: 'info',
		},
		/**
		* Especifica se o highlight vai ser estático ou animado.
		*/
		animated: {
			type: Boolean,
			default: false,
		},
		/**
		* Especifica a duração da animação.
		*/
		duration: {
			type: Number,
			default: 1,
		},
		/**
		* O tempo de espera até a animação começar.
		*/
		delay: {
			type: Number,
			default: 0,
		},
		/**
		* Define uma substring específica do texto enviada por slot que deve receber o destaque (highlight).
		*/
		highlightedText: {
			type: [String, null],
			default: null,
		},
		/**
		* Informa se o highlight vai ser Case Sensitive.
		*/
		caseSensitive: {
			type: Boolean,
			default: false,
		},
	},

	data() {
		return {
			substrings: [],
		}
	},

	computed: {
		hasSlots() {
			return !!Object.keys(this.$slots).length;
		},

		slotString() {
			return this.hasSlots ? this.$slots.default()[0].children : '';
		},

		normalizedSlotString() {
			return this.normalize(this.slotString);
		},

		normalizedHighlightText() {
			return this.normalize(this.highlightedText);
		},

		isAValidSubstring() {
			return this.normalizedSlotString.includes(this.normalizedHighlightText);
		},

		paddingResolver() {
			if (this.highlightedText) {
				return '4px 1px';
			}

			return '4px';
		},

		dynamicHighlightClass() {
			let dynamicClass = '';

			if (this.animated) {
				dynamicClass = 'highlight__container--highlighted';
			}

			switch (this.variant) {
				case 'info':
					return `${dynamicClass} highlight__container--info`;
				case 'success':
					return `${dynamicClass} highlight__container--success`;
				case 'danger':
					return `${dynamicClass} highlight__container--danger`;
				default:
					return `${dynamicClass} highlight__container--info`;
			}
		},

		dynamicStyle() {
			return {
				'--duration': `${this.duration}s`,
				'--delay': `${this.delay}s`,
			};
		},
	},

	watch: {
		highlightedText() {
			this.splitContent();
		}
	},

	mounted() {
		this.splitContent();
	},

	methods: {
		normalize(string) {
			const accentedChars = {
				'a': '[aáàâã]',
				'á': '[aáàâã]',
				'à': '[aáàâã]',
				'â': '[aáàâã]',
				'ã': '[aáàâã]',
				'e': '[eéê]',
				'é': '[eéê]',
				'ê': '[eéê]',
				'i': '[ií]',
				'í': '[ií]',
				'o': '[oóôõ]',
				'ó': '[oóôõ]',
				'ô': '[oóôõ]',
				'õ': '[oóôõ]',
				'u': '[uúü]',
				'ú': '[uúü]',
				'ü': '[uúü]',
			};

			const caseSensitiveAccentedChars = {
				'a': '[aáàâã]',
				'á': '[aáàâã]',
				'à': '[aáàâã]',
				'â': '[aáàâã]',
				'ã': '[aáàâã]',
				'e': '[eéê]',
				'é': '[eéê]',
				'ê': '[eéê]',
				'i': '[ií]',
				'í': '[ií]',
				'o': '[oóôõ]',
				'ó': '[oóôõ]',
				'ô': '[oóôõ]',
				'õ': '[oóôõ]',
				'u': '[uúü]',
				'ú': '[uúü]',
				'ü': '[uúü]',
				'A': '[AÁÀÂÃ]',
				'Á': '[AÁÀÂÃ]',
				'À': '[AÁÀÂÃ]',
				'Â': '[AÁÀÂÃ]',
				'Ã': '[AÁÀÂÃ]',
				'E': '[EÉÊ]',
				'É': '[EÉÊ]',
				'Ê': '[EÉÊ]',
				'I': '[IÍ]',
				'Í': '[IÍ]',
				'O': '[OÓÔÕ]',
				'Ó': '[OÓÔÕ]',
				'Ô': '[OÓÔÕ]',
				'Õ': '[OÓÔÕ]',
				'U': '[UÚÜ]',
				'Ú': '[UÚÜ]',
				'Ü': '[UÚÜ]',
			};

			if (this.caseSensitive) {
				return string?.replace(/[aáàâãeéêiíoóôõuúüAÁÀÂÃEÉÊIÍOÓÔÕUÚÜ]/g, match => caseSensitiveAccentedChars[match] || match);
			}

			return string?.toLowerCase().replace(/[aáàâãeéêiíoóôõuúü]/g, match => accentedChars[match] || match);
		},

		splitContent() {
			let regexFlag = this.caseSensitive ? 'g' : 'gi';
			let expression = new RegExp(`(${this.normalizedHighlightText})`, regexFlag);
			let matchResult = this.slotString.split(expression);

			this.substrings = matchResult.map((string, index) => {
				return {
					string,
					highlighted: new RegExp(`^${this.normalizedHighlightText}$`, regexFlag).test(this.caseSensitive ? string : string?.toLowerCase()),
					id: index,
				};
			});
		}
	},
};
</script>
<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

mark {
	all: unset;
}

.highlight__container {
	background-size: 200%;
	border-radius: 4px;
	padding: v-bind(paddingResolver);

	&--highlighted {
		animation: highlight var(--duration) ease-in var(--delay) backwards;
	}

	&--info {
		background-image: linear-gradient(to right, rgba($bn-300, .25) 50%, transparent 50%);
	}

	&--success {
		background-image: linear-gradient(to right, rgba($gp-300, .28) 50%, transparent 50%);
	}

	&--danger {
		background-image: linear-gradient(to right, rgba($rc-300, .22) 50%, transparent 50%);
	}
}

@keyframes highlight {
	0% {
		background-position: 100%;
	}
}
</style>
