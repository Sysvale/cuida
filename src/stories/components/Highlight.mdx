import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Highlight from '../../components/Highlight.vue';
import * as HighlightStories from './Highlight.stories';

<Meta of={HighlightStories} />

export const Template = (args) => ({
	components: { CdsHighlight: Highlight },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-highlight v-bind="args">
			O rato ROeu A RoUpA do rei de ROMA com eficácia, força e determinação.
		</cds-highlight>
	`,
});

export const DefaultSlotDemoTemplate = (args) => ({
	components: { CdsHighlight: Highlight },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-highlight>
			O rato ROeu A RoUpA do rei de ROMA
		</cds-highlight>
	`,
});

export const HighlightedTextTemplate = (args) => ({
	components: { CdsHighlight: Highlight },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-highlight highlightedText="RO">
			O rato ROeu A RoUpA do rei de ROMA
		</cds-highlight>
	`,
});

export const CaseSensitiveTemplate = (args) => ({
	components: { CdsHighlight: Highlight },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-highlight highlightedText="RO" :caseSensitive="true">
			O rato ROeu A RoUpA do rei de ROMA
		</cds-highlight>
	`,
});

# Highlight

### Highlights são componentes utilizados para destacar palavras ou frases importantes em um determinado contexto.

***

<br />

## Quando usar:

* Se deseja guiar a atenção do usuário para palavras ou frases específicas na interface (nesses cenários, Highlights animados podem ser ainda mais efetivos que os comuns).
* Estiver implementando alguma feature de *diff* e precisar indicar elementos que foram substituídos por outros.

<br />

## Quando não usar:

* Se deseja criar hierarquia na página. Para esses casos, use mecanismos de hierarquia tipográfica como peso da fonte, tamanho ou cor.
* O elemento sobre o qual o Highlight será *wrapper* não for um elemento de texto.
* Forem destacados trechos muito grandes de texto. Isso pode acabar resultando no oposto do que se espera com o uso de Highlights.

<br />

<br />

### **Uso padrão**

O comportamento padrão do `cds-highligh`, quando nenhum valor é indicado na prop `highlightedText` é o de
marcar todo o conteúdo passado por slot.

<br />

<Story of={HighlightStories.SlotDemoTemplate} />

```html
<cds-highlight>
	O rato ROeu A RoUpA do rei de ROMA
</cds-highlight>
```

<br />

<br />

### **Highlight em palavras específicas do texto**

Caso se deseje marcar apenas uma parte da string passada via slot, é possível indicar na prop
`highlightedText` a substring que receberá o destaque.

<br />

<Story of={HighlightStories.HighlightedTextTemplate_} />

```html
<cds-highlight
	highlightedText="RO"
>
	O rato ROeu A RoUpA do rei de ROMA
</cds-highlight>
```

<br />

<br />

### **Highlight em palavras específicas do texto (Case Sensitive)**

É possível indicar se as comparações feitas com a substring devem maiúsculas e minúsculas (Case Sensitive).

<br />

<Story of={HighlightStories.CaseSensitiveTemplate_} />

```html
<cds-highlight
	highlightedText="RO"
>
	O rato ROeu A RoUpA do rei de ROMA
</cds-highlight>
```

<br />

## Preview

<Canvas>
  <Story of={HighlightStories.Highlight} />
</Canvas>

<Controls />
