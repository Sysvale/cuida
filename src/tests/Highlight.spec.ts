import { describe, test, expect } from 'vitest';
import Highlight from '../components/Highlight.vue';
import { mount } from '@vue/test-utils';

describe('Highlight', () => {
	test('renders correctly', async () => {
		const wrapper = mount(Highlight, {
			props: {
				variant: 'success',
			},
			slots: {
				default: 'Texto',
			}
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
