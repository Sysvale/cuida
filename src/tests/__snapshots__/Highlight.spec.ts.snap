// Vitest Snapshot v1

exports[`Highlight > renders correctly 1`] = `
"<mark data-v-eddeabb9=\\"\\" class=\\"highlight__container highlight__container--success\\" style=\\"--duration: 1s; --delay: 0s;\\">
  <!-- @slot Slot usado para especificar o texto que receberÃ¡ o highlight. -->Texto
</mark>"
`;
